### PR TITLE
Epic 8438 : Assay Data Tutorials

### PR DESCRIPTION
**Pull request by labkey-bpatel**
#### Rationale
Allow adding of assay tutorials to help users understand the assay data.
Spec [here](https://docs.google.com/document/d/1MtSvDRxyOHmpg07XQZdWB4Etbwza0nwhx80s3ZMdMi4/edit#heading=h.s8qmk21y0t38).

#### Changes
* Schema changes to add assay tutorials links and downloadable documents
* ETL updates
* Assay tutorial view, model, and store changes
* Show tutorial in a popup modal
* Embed tutorial display name in a thumbnail

### CODE REVIEW

File: src/org/labkey/cds/CDSController.java [hunk 1]
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
...
+            String basePath = CDSManager.get().getAssayDocumentPath(getContainer());
+
+            if (StringUtils.isBlank(basePath))
+            {
+                errors.reject(ERROR_MSG, "Assay document location not specified.");
+            }
+
+            if (!errors.hasErrors() && !StringUtils.isBlank(filename))
+            {
+                if (isFileValid(errors, response, filename, basePath))
```
Comment (XingY) [inline src/org/labkey/cds/CDSController.java]: The function name is very confusing, maybe rename to downloadFile?

File: src/org/labkey/cds/CDSController.java [hunk 2]
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
...
+    {
+        @Override
+        public void addNavTrail(NavTree root)
+        {
+        }
+
+        @Override
+        public void validate(StudyDocumentForm form, BindException errors)
+        {
+            if (null == form.getDocumentId())
```
Comment (XingY) [inline src/org/labkey/cds/CDSController.java]: It doesn't seem to me documentId is used by this action?

File: webapp/Connector/src/app/view/module/AssayTutorial.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    },
+
+    getListData : function () {
+        return this.initialConfig.data.model.data;
+    },
+
+    hasContent : function() {
+        var d = this.getListData();
+        return d["assayTutorialDocuments"].length > 0 || d["assayTutorialLinks"].length > 0;
```
Comment (XingY) [inline webapp/Connector/src/app/view/module/AssayTutorial.js]: can use d["assayTutorialDocuments"]?.length > 0 || d["assayTutorialLinks"]?.length > 0 to check for null

File: src/org/labkey/cds/CDSController.java [hunk 4]
```diff
@@ -1126,6 +1126,52 @@ public Object execute(StudyDocumentForm form, BindException errors)
...
+                errors.reject(ERROR_MSG, "documentId parameter required");
+            if (null == form.getFilename())
```
Comment (XingY) [inline src/org/labkey/cds/CDSController.java]: Can use StringUtils.isBlank here

File: webapp/Connector/src/app/view/module/AssayTutorial.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
+Ext.define('Connector.view.module.AssayTutorial', {
+    xtype: 'app.module.assaytutorial',
+
+    extend : 'Connector.view.module.ShowList',
+
+    plugins : ['documentvalidation'],
+
+    statics : {
+        openPopupModal : function(videoUrl) {
```
Comment (XingY) [inline webapp/Connector/src/app/view/module/AssayTutorial.js:9]: Seems this is similar to to the win obj in Started.js maybe can reuse?

File: webapp/Connector/src/app/view/module/AssayTutorial.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                                    '</div>',
+                                '</div>',
+                            '</td></tr>',
+                        '</tpl>',
+                    '</table>',
+                '</tpl>',
+            '<tpl if="assayTutorialDocuments && assayTutorialDocuments.length &gt; 0">',
+                '<table class="learn-study-info">',
+                    '<tpl for="assayTutorialDocuments">',
+                        '<tr><td>',
```
Comment (XingY) [inline webapp/Connector/src/app/view/module/AssayTutorial.js]: check isLinkValid?

**Review by XingY — COMMENTED** [review]
Just a few comments.

File: test/sampledata/dataspace/cdsimport/Document.txt [hunk 7]
```diff
@@ -175,4 +175,9 @@ document_id	label	filename	document_type	assay_identifier
 204	ILLUMINA 454-X	study_documents/ILLUMINA_454_X.tsv	Non-Integrated Assay	ILLUMINA 454-X
 300	Publication Datasets - Monoclonal antibodies	study_documents/alvac.pdf	Publication Data	
 301	Analysis Data - T Cell Immune Response	study_documents/t_cell_data.zip	Publication Data	
-302	Publication Datasets	study_documents/t_cell_ir.tsv	Publication Data	
\ No newline at end of file
+302	Publication Datasets	study_documents/t_cell_ir.tsv	Publication Data
+303	NAB tutorial A	assay_documents/nab_tutorial_a.pdf	Assay Tutorial Document	NAB A3R5
+304	NAB tutorial B	assay_documents/nab_tutorial_b.docx	Assay Tutorial Document	NAB A3R5
+306	Intracellular Cytokine Staining tutorial	assay_documents/ics_tutorial.pdf	Assay Tutorial Document	ICS
+305	NAB YouTube link 1		Assay Tutorial Link	NAB A3R5	https://www.youtube.com/embed/Q3BPWNYyv14	Introduction to NAb (Neutralizing antibody assay)
```
Comment (XingY) [inline test/sampledata/dataspace/cdsimport/Document.txt:182]: :thumbsup: good job Chef Kobe

Comment (labkey-bpatel) [reply test/sampledata/dataspace/cdsimport/Document.txt:182]: Haha! indeed!

File: webapp/Connector/src/app/view/module/AssayTutorial.js (hunk 6 above)
Comment (labkey-bpatel) [reply webapp/Connector/src/app/view/module/AssayTutorial.js]: done

File: webapp/Connector/src/app/view/module/AssayTutorial.js (hunk 5 above)
Comment (labkey-bpatel) [reply webapp/Connector/src/app/view/module/AssayTutorial.js:9]: i had tried reusing, however it just did not work as expected, so went this route

File: src/org/labkey/cds/CDSController.java (hunk 4 above)
Comment (labkey-bpatel) [reply src/org/labkey/cds/CDSController.java]: done

File: webapp/Connector/src/app/view/module/AssayTutorial.js (hunk 3 above)
Comment (labkey-bpatel) [reply webapp/Connector/src/app/view/module/AssayTutorial.js]: done

File: src/org/labkey/cds/CDSController.java (hunk 2 above)
Comment (labkey-bpatel) [reply src/org/labkey/cds/CDSController.java]: correct, now removed

File: src/org/labkey/cds/CDSController.java (hunk 1 above)
Comment (labkey-bpatel) [reply src/org/labkey/cds/CDSController.java]: i agree, now renamed

**Review by XingY — APPROVED** [review]

File: webapp/Connector/src/app/view/module/AssayTutorial.js (hunk 3 above)
Comment (labkey-bpatel) [reply webapp/Connector/src/app/view/module/AssayTutorial.js]: prod build did not like ?. syntax, so this is now changed to:
(d["assayTutorialDocuments"] && d["assayTutorialDocuments"].length > 0) || (d["assayTutorialLinks"] && d["assayTutorialLinks"].length > 0)

**Pull request merged**